include/RED4ext/SharedPtr.hpp

```diff
@@ -0,0 +1,377 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+#include <Windows.h>
+
+#include <RED4ext/Addresses.hpp>
+#include <RED4ext/Common.hpp>
+#include <RED4ext/Relocation.hpp>
+#include <RED4ext/Memory/Utils.hpp>
+
+namespace RED4ext
+{
+struct RefCnt
+{
+    using AllocatorType = Memory::RefCountAllocator;
+
+    constexpr RefCnt() noexcept
+        : strongRefs(1)
+        , weakRefs(1)
+    {
+    }
+
+    void IncRef()
+    {
+        InterlockedIncrement(&strongRefs);
+    }
+
+    // Returns true if the strong refs count has been successfully incremented.
+    bool IncRefIfNotZero()
+    {
+        uint32_t uses = strongRefs;
+        while (uses != 0)
+        {
+            const uint32_t oldUses = InterlockedCompareExchange(&strongRefs, uses + 1, uses);
+            if (oldUses == uses)
+                return true;
+
+            uses = oldUses;
+        }
+
+        return false;
+    }
+
+    // Returns true if the strong refs count reached zero.
+    bool DecRef()
+    {
+        return InterlockedExchangeAdd(&strongRefs, static_cast<uint32_t>(-1)) == 1;
+    }
+
+    void IncWeakRef()
+    {
+        InterlockedIncrement(&weakRefs);
+    }
+
+    volatile uint32_t strongRefs;
+    volatile uint32_t weakRefs;
+};
+RED4EXT_ASSERT_SIZE(RefCnt, 0x8);
+
+template<typename T>
+class SharedPtrBase
+{
+public:
+    SharedPtrBase(const SharedPtrBase&) = delete;
+
+    ~SharedPtrBase() = default;
+
+    SharedPtrBase& operator=(const SharedPtrBase&) = delete;
```

Comment:
Why deleted?

Move this one before the dtor.

Reply:
Not sure, it's straight from previous `HandleBase`.

Reply:
I would keep it honestly. A copy for shared ptr is good imo.

Reply:
All copy/move ctor/ops are declared for final types `SharedPtr` and `Handle`. 
This looks more like a precaution: to not use ctor/ops on base type.

Reply:
I was thinking to make them protected so we can move or copy SharedPtr around. If we make them delete we can't move them even from SharedPtr class (if I remember right).

Reply:
Checked. Operators in children (`Handle`, etc.) work.

Reply:
Nice! Then that's good.